pkg/conditions/error.go

```diff
@@ -0,0 +1,24 @@
+package conditions
+
+import (
+	"github.com/giantswarm/microerror"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+)
+
+var unexpectedConditionStatusError = &microerror.Error{
+	Kind: "UnexpectedConditionStatus",
+}
+
+func UnexpectedConditionStatusError(cr CR, t capi.ConditionType) error {
+	c := capiconditions.Get(cr, t)
+
+	return microerror.Maskf(
```

Comment:
I'm not sure if this is a good idea. The return stack trace looses the real place of error origin, when used e.g. like below:
```go
	// Let's make sure that the condition status is set to a supported value.
	if conditions.IsUnexpected(machinePool, aeconditions.UpgradingCondition) {
		return conditions.UnexpectedConditionStatusError(machinePool, aeconditions.UpgradingCondition)
	}
```

The last two entries in returned stack trace would be the caller of function from where this snippet is from, given it used `microerror.Mask()` and then the line from this error constructor.

Maybe this could be changed so that this function is reduced to:
```go
func UnexpectedConditionStatusErrorMsg(cr CR, t capi.ConditionType) string {
	c := capiconditions.Get(cr, t)
	return fmt.Sprintf("Unexpected status for condition %s, got %s", t, c.Status)
}
```

And the call site would be:
```go
	// Let's make sure that the condition status is set to a supported value.
	if conditions.IsUnexpected(machinePool, aeconditions.UpgradingCondition) {
		return microerror.Maskf(conditions.UnexpectedConditionStatusError, conditions.UnexpectedConditionStatusErrorMsg(machinePool, aeconditions.UpgradingCondition))
	}
```

That is not so beautiful, I admit, but it would preserve the expected stack trace.

Reply:
Good point :+1: